rfcs/51-animation-composition.md

```diff
@@ -0,0 +1,450 @@
+
+```rust
+struct Track<T> {
+  curves: Vec<Option<Arc<dyn Curve<T>>>>,
+}
+```
+
+The individual curves stored within are indexed by `ClipId`, and each pointer to
+a curve is non-None if and only if the original clip had a curve for the
+corresponding property.
+
+A type erased version of `Track<T>` will be needed to store the track in a
+concrete map. The generic type parameter of the sample function will be used to
+downcast the type erased version into the generic version.
+
+As mentioned in the prior primitives RFC, there will likely need to be a separate
+clip storage dedicated to storing specialized `Transform` curves for the purposes
+of skeletal animation that avoids both the runtime type casting and the dynamic
+dispatch of `Arc<dyn Curve<T>>`.
+
+### Graph Evaluation
+```rust
+struct Influence {
+    weight: f32,
+    time: f32,
+}
+
+struct GraphInfluences {
+    influence: Vec<Influence> // Indexed by BoneId
+}
+```
+
+To remove the need to evaluate the graph every time a property is sampled, an
+influences map is constructed based on the current state of the graph. This is
+effectively a `Vec<f32>` indexed on `ClipId` mapping clips and their respective
+*cumulative* weights. To avoid constant reallocation, this influences map is
+stored as a companion component and is cleared before every evaluation.
+
+During evaluation, the graph is traversed along every connected edge and weights
+are multiplicatively propagated downward. For example, if we have a simple graph
+`A -> B -> C`, where A is the root and C is the final clip and the edges `A -> B`
+and `B -> C` have the weights 0.5 and 0.25 respectively, the final weight for
+clip C is 0.125. If multiple paths to a end clip are present, the final weight
+for the clip is the sum of all path's weights.
+
+After traversing the graph, the weights of all active inputs are clamped and normalized.
+
+By default, a dedicated system will run before sampling every app tick that
+evaluates every changed graph, relying on change detection to ensure that any
+mutation to the graph results in a change in the influences. This behavior can be
+disabled via a config option, requiring users to manually evaluate the graph
+instead.
+
+### Binding
+
+```rust
+struct BoneId(usize);
+
+#[derive(PartialEq, Eq, ParitalOrd, Ord)]
+struct EntityPath {
+    path: Vec<Name>,
+}
+
+struct GraphBindings {
+    paths: BTreeMap<EntityPath, BoneId>,
+    bones: Vec<Bone>,
+}
+
+struct Bone {
+    properties: BTreeMap<FieldPath, Track>,
+}
+
+#[derive(Resource)]
+struct BoneBindings(SparseSet<Entity, SmallVec<[(Entity, BoneId), 2]>>);
+```
+
+After graph evalution is building a global `BoneBindings` resource. This is a
+newtype over a mapping between an animated entity (the key), and the
+`AnimationGraph`s and corresponding `BoneId`s the entity is bound to.
+
+`BoneBindings` allows us to safely mutate all animated entities in parallel. This
```

Comment:
> Can we instead utilize the DAG nature of the entity graph (and the fact that AnimationGraph operates on a child hierarchy of the DAG) to animate AnimationGraphs in parallel by starting at "the top" of the entity hierarchy (filtered down to entities with AnimationGraph) and "animating down"? Peer animation graphs could never reference the same entity. A given AnimationGraph should only need to mutate a given Bone entity's component once at the end of the graph execution.

This would require us to assume that graphs do not have overlapping trees. Consider the following hierarchy:

```
        A
      /     \
    B       C
           /    \
         D      E
```

If both A and C have animation graphs targeting C or D, they will still alias.

Unless we enforce that only one animation graph can work in a given hierarchy and/or that animation graphs only live on root entities, this likely is intractable.

Reply:
> Unless we enforce that only one animation graph can work in a given hierarchy and/or that animation graphs only live on root entities, this likely is impossible to get it to parallelize without explicit deduplication.

Under these circumstances, I think it is safe to parallelize everything within a given depth of the hierarchy in parallel, as entities at a given depth are guaranteed to not contain each other in their DAG, which is the pattern I was trying to get at. In the graph above: B and C graphs could be animated in parallel.

Given that a graph will often be animating X "things" at once across many entities (bones, components, etc), and those things are parallelizable, and that we can parallelize graphs at the sibling level, it seems like that already introduces a reasonably high chance of core saturation, especially given the high likelihood of "sibling-ness" for animated things.

>  I also generally want us to avoid more full hierarchy traversals if we can.

Agreed. But when compared with the cost of constructing per-entity access lists each frame, I might still bet on simple hierarchy traversals winning the perf game.

Reply:
>  B and C graphs could be animated in parallel

Important to also call out that B descendants with graphs and C descendants with graphs can also run in parallel with each other, regardless of depth (provided C descendants honor the rules outlined above).

Reply:
The fact that A can't run in parallel with C doesn't seem like a huge loss under those circumstances. A less common pattern I think, given that graphs will generally be orchestrating distinct logical units.

Reply:
> I think it is safe to parallelize everything within a given depth of the hierarchy in parallel

I don't think we have a solid way of querying for all entities at a specific hierarchy level, unless we carefully do a full hierarchy breadth-first search, which I still think we should avoid doing single threaded. What we could do instad is just parallelize on all graphs regardless of level, but predicate any actual animation with an ancestor query to ensure there is no higher level graph that might alias against it. If most animation graphs are root or close to root level, these ancestor queries should be fairly cheap. However, this would force it to parallelize only at the graph level: the hierarchy below the graph and the sampling application for each graph must be single threaded. This means that any graph with a large animated hierarchy under it are inevitably going to be a straggler. If there is a way for us to do parallel deduplication without allocation, we can avoid these stragglers and ensure better work distribution across threads.

Reply:
Swapped up how we're supposed to do parallelization, which maps to the current parallelization strategy introduced in https://github.com/bevyengine/bevy/pull/6785.